<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link href="./library/css/bootstrap.min.css" rel="stylesheet">
</head>
<body background="./library/image/back.png">
    <div class="container">
		<div class="row">
			<div class="col-md-3">
				<br>
				<a href="http://www.genenergy.cn/"><img src="./library/image/Rlogo.jpg"/></a>
			</div>
			<div class="col-md-9">
				<h2 align="center">10X Genomics 单细胞测序数据分析报告</h2>
				<h4 align="center">Custom:Custom &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Contract:Contract</h4>
			</div>   
		</div> 
		<hr/>
		<div class="row">
			<div class="col-md-3" >
				<div class="navbar navbar-default" role="navigation">
					<ul class="nav">
						<li><a href="01_Index.html">Introduction(简介)</a></li>
						<li><a href="02_QC.html">QC(测序数据质量统计)</a></li>					
						<li><a href="03_Cluster.html">Cluster(细胞聚类)</a></li>
						<li><a href="04_CellMarker.html">CellMarker(细胞标记基因分析)</a></li>
						<li><a href="05_TrajectoryAnalysis.html">Trajectory(拟时轨迹分析)</a></li>
						<li><a href="06_FUN.html">FUN(细胞标记基因功能分析)</a></li>
						<li><a href="07_Datasoft.html">DBSoft(数据库和软件)</a></li>
						<li><a href="08_Reference.html">Refs(参考文献)</a></li>
						<li><a href="09_Help.html">F&Q(帮助)</a></li>
					</ul>
				</div>	  
			</div>
			<div class="col-md-9">	
                <h3>参考文献</h3>		    
				<ol>
					<li>Cell ranger :http://support.10xgenomics.com/single-cell/software/overview/welcome</li>
					<li>Visualizing Data using t-SNE. Laurens van der Maaten et al. Journal of Machine Learning Research 1 (2008) 1-48. </li>
					<li>New clustering proposals for Twitter based on the Louvain algorithm, Master’s Thesis (UniversidadNacional de Educacio´ n a Distancia). Matutano, J.C., Unanue, R.M., and Avanzada, M.U.E.I. (2014). </li>
					<li>STRING v9.1: protein-protein interaction networks, with increased coverage and integration. Andrea Franceschini, Damian Szklarczyk, Sune Frankild, Michael Kuhn, Milan Simonovic, Alexander Roth, Jianyi Lin, Pablo Minguez, Peer Bork, Christian von Mering and Lars J. Jensen. Nucleic Acids Research. 2013. 41: D808–D815. doi:10.1093/nar/gks1094.</li>
                    <li>Small-sample estimation of negative binomial dispersion, with applications to SAGE data. Robinson, M. D. & Smyth, G. K. Biostatistics 9, 321–332 (2007).</li>
					<li>Shrinkage estimation of dispersion in Negative Binomial models for RNA-seq experiments with small sample size. Yu, D., Huber, W. & Vitek, O. Bioinformatics 29, 1275–1282 (2013).</li>
					<li>Data exploration, quality control and testing in single-cell qPCR-based gene expression experiments. Andrew McDavid, et al. Bioinformatics. 2013.</li>
					<li>Reversed graph embedding resolves complex single-cell developmental trajectories. Xiaojie Qiu, et al. Nature Methods. 2017. 14(10):979–982.</li>
					</ol>
				
			</div>
		</div>	
    </div>
	<div id="footer" class="container">
		<nav class="navbar navbar-default">
			<div class="navbar-inner navbar-content-center">
			</div>
		</nav>
	</div>
</body>
</html>
